frontend/catalyst/pennylane_extensions.py

```diff
@@ -1235,33 +1338,78 @@ def _call_with_quantum_ctx(self, ctx):
         in_classical_tracers = self.preds
         regions: List[HybridOpRegion] = []
 
-        out_trees, out_avals = [], []
-        for branch in self.branch_fns + [self.otherwise_fn]:
+        in_sigs, out_sigs = [], []
+        for branch_fn in self.branch_fns + [self.otherwise_fn]:
             quantum_tape = QuantumTape()
+            wfun, in_sig, out_sig = deduce_signatures(
+                branch_fn, [], {}, expansion_strategy=self.expansion_strategy
+            )
+            assert len(in_sig.in_type) == 0
             with EvaluationContext.frame_tracing_context(ctx) as inner_trace:
-                wffa, _, _, out_tree = deduce_avals(branch, [], {})
                 with QueuingManager.stop_recording(), quantum_tape:
-                    res_classical_tracers = [inner_trace.full_raise(t) for t in wffa.call_wrapped()]
-            regions.append(HybridOpRegion(inner_trace, quantum_tape, [], res_classical_tracers))
-            out_trees.append(out_tree())
-            out_avals.append(res_classical_tracers)
-
-        _check_cond_same_shapes(out_trees, out_avals)
-        res_avals = list(map(shaped_abstractify, res_classical_tracers))
-        out_classical_tracers = [new_inner_tracer(outer_trace, aval) for aval in res_avals]
-        Cond(in_classical_tracers, out_classical_tracers, regions)
-        return tree_unflatten(out_tree(), out_classical_tracers)
-
-    def _call_with_classical_ctx(self):
-        args, args_tree = tree_flatten([])
-        args_avals = tuple(map(_abstractify, args))
-        branch_jaxprs, consts, out_trees = initial_style_jaxprs_with_common_consts1(
-            (*self.branch_fns, self.otherwise_fn), args_tree, args_avals, "cond"
+                    res_classical_tracers = [inner_trace.full_raise(t) for t in wfun.call_wrapped()]
+
+            regions.append(
+                HybridOpRegion(
+                    inner_trace,
+                    quantum_tape,
+                    [],
+                    collapse(out_sig.out_type(), res_classical_tracers),
+                )
+            )
+            in_sigs.append(in_sig)
+            out_sigs.append(out_sig)
+
+        _check_cond_same_shapes([s.out_tree() for s in out_sigs])
+        out_tree = out_sigs[-1].out_tree()
+        all_consts = [s.out_consts() for s in out_sigs]
+        out_types = [s.out_type() for s in out_sigs]
+        # FIXME: Creating output tracers like this is not quite cocrect. We need to calculate the
+        # unified result type first and only then use it as the output type of this primitive.
+        out_type2 = out_types[-1]
```

Comment:
This is the issue I discovered several days ago. It was always present, but we didn't hit it because I believe we don't support conditionals in user-level quantum transformations (according to Erick). Here is the [issue](https://github.com/PennyLaneAI/catalyst/issues/403) that I created.

Reply:
Thanks!